File: misc/wrapgen/static/apply.jl
# SPDX-License-Identifier: Apache-2.0
# STATIC hand-written special -- copied verbatim by generate.sh.
#
# low-level operator multiplication routine
export apply


function apply(ops::OpSum,
               bin::Block, vin::Vector{Float64},
               bout::Block, vout::Vector{Float64})
    if (length(vin) != length(bin))
        error("Input vector does not have the same size as input Block")
    elseif (length(vout) != length(bout))
        error("Output vector does not have the same size as output Block")
    end
    
    GC.@preserve vin vout begin
        fun_apply(ops.cxx_object, bin.cxx_object, pointer(vin),
                  bout.cxx_object, pointer(vout))
    end
end
function apply(ops::OpSum,
               bin::Block, vin::Vector{ComplexF64},
               bout::Block, vout::Vector{ComplexF64})
    if (length(vin) != length(bin))
        error("Input vector does not have the same size as input Block")
    elseif (length(vout) != length(bout))
        error("Output vector does not have the same size as output Block")
    end
    GC.@preserve vin vout begin
        fun_applyC(ops.cxx_object, bin.cxx_object, pointer(vin),
                                 bout.cxx_object, pointer(vout))
    end
end
function apply(ops::OpSum,
               bin::Block, min::Matrix{Float64},
               bout::Block, mout::Matrix{Float64})
    if (size(min, 1) != length(bin))
        error("Input matrix does not have the same number of rows as the size of the input Block")
    elseif (size(mout, 1) != length(bout))
        error("Output matrix does not have the same number of rows as the size of the output Block")
    elseif size(min, 2) != size(mout, 2)
        error("Input and output matrices need to have the same number of columns")
    end
    GC.@preserve min mout begin
        fun_apply(ops.cxx_object, bin.cxx_object, pointer(min),
                  bout.cxx_object, pointer(mout), size(min, 2))
    end
end
function apply(ops::OpSum,
               bin::Block, min::Matrix{ComplexF64},
               bout::Block, mout::Matrix{ComplexF64})
    if (size(min, 1) != length(bin))
        error("Input matrix does not have the same number of rows as the size of the input Block")
    elseif (size(mout, 1) != length(bout))
        error("Output matrix does not have the same number of rows as the size of the output Block")
    elseif size(min, 2) != size(mout, 2)
        error("Input and output matrices need to have the same number of columns")
    end
    GC.@preserve min mout begin
        fun_applyC(ops.cxx_object, bin.cxx_object, pointer(min),
                   bout.cxx_object, pointer(mout), size(min, 2))
    end
end

# Op / Monomial: a one-term OpSum, so reuse the zero-copy OpSum path.
apply(op::Op, bin::Block, vin, bout::Block, vout) =
    apply(OpSum(op), bin, vin, bout, vout)
apply(op::Monomial, bin::Block, vin, bout::Block, vout) =
    apply(OpSum(op), bin, vin, bout, vout)

File: julia/src/apply.cpp
// SPDX-License-Identifier: Apache-2.0
// STATIC hand-written special -- copied verbatim by generate.sh.
//
// Dense matrix builder: fills a caller-allocated, Julia-owned column-major
// buffer of size size(out) x dim(in). The output block is inferred via
// block(ops, in) (so Julia never round-trips the Block variant). Templated on
// coeff_t, so instantiated here per block. Sparse builders live in sparse.cpp.

#include <julia/src/xdiagjl.hpp>
#include <xdiag/armadillo.hpp>

namespace xdiag::julia {

template <typename block_t> static void register_apply(jlcxx::Module &mod) {

  // Vector variants
  mod.method("fun_apply", [](OpSum const &ops, block_t const &bin, double *pin,
                             block_t const &bout, double *pout) {
    int64_t sin = bin.size();
    int64_t sout = bout.size();
    arma::vec vin(pin, sin, /*copy_aux_mem*/ false, /*strict*/ false);
    arma::vec vout(pout, sout, /*copy_aux_mem*/ false, /*strict*/ false);
    JULIA_XDIAG_CALL_VOID((apply(ops, bin, vin, bout, vout)));
  });

  mod.method("fun_applyC", [](OpSum const &ops, block_t const &bin,
                              complex *pin, block_t const &bout,
                              complex *pout) {
    int64_t sin = bin.size();
    int64_t sout = bout.size();
    arma::cx_vec vin(pin, sin, /*copy_aux_mem*/ false, /*strict*/ false);
    arma::cx_vec vout(pout, sout, /*copy_aux_mem*/ false, /*strict*/ false);
    JULIA_XDIAG_CALL_VOID((apply(ops, bin, vin, bout, vout)));
  });

  // Matrix variants
  mod.method("fun_apply", [](OpSum const &ops, block_t const &bin, double *pin,
                             block_t const &bout, double *pout, int64_t ncols) {
    int64_t sin = bin.size();
    int64_t sout = bout.size();
    arma::mat min(pin, sin, ncols, /*copy_aux_mem*/ false, /*strict*/ false);
    arma::mat mout(pout, sout, ncols, /*copy_aux_mem*/ false, /*strict*/ false);
    JULIA_XDIAG_CALL_VOID((apply(ops, bin, min, bout, mout)));
  });

  mod.method("fun_applyC", [](OpSum const &ops, block_t const &bin,
                              complex *pin, block_t const &bout, complex *pout,
                              int64_t ncols) {
    int64_t sin = bin.size();
    int64_t sout = bout.size();
    arma::cx_mat min(pin, sin, ncols, /*copy_aux_mem*/ false, /*strict*/ false);
    arma::cx_mat mout(pout, sout, ncols, /*copy_aux_mem*/ false,
                      /*strict*/ false);
    JULIA_XDIAG_CALL_VOID((apply(ops, bin, min, bout, mout)));
  });
}

void define_apply(jlcxx::Module &mod) {
  using namespace xdiag;
  register_apply<Spinhalf>(mod);
  register_apply<tJ>(mod);
  register_apply<Electron>(mod);
  register_apply<Boson>(mod);
  register_apply<Fermion>(mod);
}

} // namespace xdiag::julia
